main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+  * as well as for useless patterns
+  *
+  * A pattern is useless if:
+  * A pattern match is exhaustive if:
+  */
+object PatternExhaustiveness extends Phase[TypedAst.Root, TypedAst.Root] {
+
+  /**
+    * Returns an error message if a pattern match is not exhaustive
+    */
+  def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = {
+    if (flix.options.debug) {
+      // Can probably do options stuff here
+    }
+    implicit val _ = flix.genSym
+    val defns = seqM(root.definitions.map { case (_, v) => Definition.CheckPats(root, v) })
+
+    defns.map ( _ => root)
+  }
+
+  object Definition {
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Declaration.BoundedLattice)(implicit genSym: GenSym): Validation[TypedAst.Declaration.BoundedLattice, CompilationError] = tast match {
+      case TypedAst.Declaration.BoundedLattice(tpe, bot, top, leq, lub, glb, loc) =>
+        for {
+          _bot <- Expression.CheckPats(root, bot)
+          _top <- Expression.CheckPats(root, top)
+          _leq <- Expression.CheckPats(root, leq)
+          _lub <- Expression.CheckPats(root, lub)
+          _glb <- Expression.CheckPats(root, glb)
+        } yield TypedAst.Declaration.BoundedLattice(tpe, _bot, _top, _leq, _lub, _glb, loc)
+    }
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Declaration.Definition)(implicit genSym: GenSym): Validation[TypedAst.Declaration.Definition, CompilationError] = tast match {
+      case TypedAst.Declaration.Definition(doc, ann, sym, tparams, fparams, exp, tpe, loc)  =>
+        for {
+        _exp <- Expression.CheckPats(root, exp)
+        } yield TypedAst.Declaration.Definition(doc, ann, sym, tparams, fparams, _exp, tpe, loc)
+    }
+
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Declaration.Index)(implicit genSym: GenSym): TypedAst.Declaration.Index = tast
+  }
+
+  object Expression {
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Expression)(implicit genSym: GenSym): Validation[TypedAst.Expression, CompilationError] = tast match {
+      case TypedAst.Expression.Match(exp, rules, tpe, loc) =>
+        for  {
+          _rules <- CheckRules(root, exp, rules)
+        } yield TypedAst.Expression.Match(exp, _rules, tpe, loc)
+
+
+      case a => Success(a, Stream.empty)
+    }
+
+    def CheckRules(root: TypedAst.Root, exp: TypedAst.Expression, rules: List[TypedAst.MatchRule]): Validation[List[TypedAst.MatchRule], CompilationError] = {
+      FindNonMatchingPat(root, exp, rules.map(r => List(r.pat)), 1) match {
+        case Right(a) => Success(rules, Stream.empty)
+        case Left((c, rule)) => Failure(ExhaustiveMatchError(formatRule(rule), rule.head.loc.source) #:: Stream.empty);
+      }
+    }
+
+    /**
+      * Format a pattern for printing so the user knows what they missed
+      *
+      * @param rule The rule to print
+      * @return The formatted string
+      */
+    def formatRule(rule: List[TypedAst.Pattern]): String = {
```

Comment:
If this is strictly for pretty printing, then let us call it `prettyPrint` or at least something that includes those words.